lib/id3c/cli/command/etl/fhir.py

```diff
@@ -0,0 +1,712 @@
+        *tract_hierarchy*.
+        """
+        # If we have an address identifier ("household id"), we upsert a
+        # location record for it.  Addresses are not reasonably enumerable, so
+        # we don't require they exist.
+        scale = 'address'
+        address_identifier = identifier(location, f"{INTERNAL_SYSTEM}/location/{scale}")
+
+        if not address_identifier:
+            return None
+
+        return upsert_location(db,
+            scale       = scale,
+            identifier  = address_identifier,
+            hierarchy   = tract_hierarchy)
+
+
+    code = location_code(location)
+    relation = location_relation(code)
+
+    try:
+        parent_location = location.partOf.resolved(Location)
+        tract_hierarchy = get_tract_hierarchy(parent_location)  # TODO do we only want parent hierarchy?
+
+    except AttributeError:
+        LOG.info(f"No parent location found for Location with code «{code}», " + \
+            f"relation «{relation}».")
+        tract_hierarchy = None
+
+    address = process_address(location, tract_hierarchy)
+
+    if not (tract_hierarchy or address):
+        LOG.warning(f"No tract or address location available for «{relation}»")
+        return
+
+    upsert_encounter_location(db,
+        encounter_id    = encounter_id,
+        relation        = relation,
+        location_id     = address.id)
+
+
+def location_code(location: Location) -> str:
+    """
+    Given a *Location* Resource, returns the matching system code for our
+    internal, location-relation system.
+    """
+    codes: List[str] = []
+    codes += list(map(lambda t: matching_system_code(t, LOCATION_RELATION_SYSTEM), location.type))
+
+    if not codes:
+        return None
+
+    unique_codes = list(set(codes))
+    if len(unique_codes) > 1:
+        raise Exception(f"Expected only one Location code. Instead received {unique_codes}.")
+
+    return unique_codes[0]
+
+
+def process_sample(db: DatabaseSession, identifier: str) -> Any:
+    """ Given an *identifier*, returns its matching sample from ID3C. """
+
+    sample = find_sample(db, identifier)
+
+    if not sample:
+        raise SampleNotFoundError(f"No sample with identifier «{identifier}» found.")
+
+    return sample
+
+
+def process_presence_absence_tests(db: DatabaseSession, report: DiagnosticReport,
+    sample_id: int, barcode: str):
+    """
+    Given a *report* containing presence-absence test results, upserts them to
+    ID3C, attaching a sample and target ID.
+    """
+    for result in report.result:
+        observation = result.resolved(Observation)
+
+        target_identifier = matching_system_code(observation.code, TARGET_SYSTEM)
+        # Most of the time we expect to see existing targets so a
+        # select-first approach makes the most sense to avoid useless
+        # updates.
+        target = find_or_create_target(db,
+            identifier  = target_identifier,
+            control     = False)
```

Comment:
This currently uses the SNOMED code as the `target.identifier`. Is this the behavior that we want, or do we want to map the SNOMED code to human readable identifiers?

Reply:
I think best to use the SNOMED code as the `target.identifier`, but importantly, it needs to be a fully-qualified SNOMED URL, not just the numeric `code` value (which is meaningless without qualifying the system it's from). https://www.hl7.org/fhir/snomedct.html#4.3.1.0.6 describes how to form those URLs, which is basically:

```python
system_concept_url = {
    "http://snomed.info/sct": "http://snomed.info/id/",
}
return urljoin(system_concept_url[concept.system], concept.code)
```

Reply:
This seems like a blocker for deploy tomorrow.